wiseconnect: Add Bluetooth HCI driver

silabs-TiborL opened the pull request.

silabs-TiborL committed:
wiseconnect: Add Bluetooth HCI driver

Origin: Silicon Labs WiSeConnect SDK
License: MSLA
URL: https://github.com/siliconlabs/wiseconnect
Commit: e97a0ed00ddda347a8a39e8276f470e1c5fea469
Version: 3.3
Purpose: Add Bluetooth support for SiWx917

Signed-off-by: Tibor Laczko <tibor.laczko@silabs.com>

asmellby commented on wiseconnect/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c:
```diff
@@ -783,7 +783,8 @@ sl_status_t sl_si91x_gpio_driver_enable_pad_selection(uint8_t gpio_padnum)
   // Check if the GPIO pin number exceeds the maximum allowed value.
-  if (gpio_num >= HOST_PAD_MIN && gpio_num <= HOST_PAD_MAX) {
+  //if (gpio_num >= HOST_PAD_MIN && gpio_num <= HOST_PAD_MAX) { FIXME SDK bug
+  if (gpio_num < HOST_PAD_MIN || gpio_num > HOST_PAD_MAX) {
```
This is fixed in a different way in Wiseconnect 3.3.1. It would be cleaner if we apply the exact same change as upstream for now:

```
  if (!(gpio_num >= HOST_PAD_MIN && gpio_num <= HOST_PAD_MAX)) {
```

Also, patches should preferably be done in separate commits from the ones introducing unmodified code from upstream.

We also want to upgrade to 3.3.2 soon, but let's keep that separate from this PR.

silabs-TiborL replied:
I did not want to add this to this PR, will revert.

asmellby reviewed (commented):
Please make sure that your commit message contains origin information including the git hash from which the content was imported. See https://docs.zephyrproject.org/latest/contribute/guidelines.html#identifying-contribution-origin for instructions, or any of the previous commits introducing HAL content.

jhedberg commented:
@silabs-TiborL you need to rebase, since #7 introduced changes to `CMakeLists.txt` that conflicts with this one.

silabs-TiborL committed:
wiseconnect: BLE and WiFi related SDK patches

 - SDK typo fixes to pass the Zephyr compliance check.
 - Fix sscanf compiler warning.
 - Remove the need of an external ble_config.h file (common config was extended with 3 default macro).

Signed-off-by: Tibor Laczko <tibor.laczko@silabs.com>

jhedberg commented on wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h:
```diff
@@ -302,7 +302,7 @@
 /// @note Bit 30 is reserved
 
 /// Extention valid to use Extended custom feature bitmap
-#define SL_SI91X_CUSTOM_FEAT_EXTENTION_VALID BIT(31)
+#define SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID BIT(31)
```
Is this purely a downstream change in this tree, or was the same fix also submitted to the upstream wiseconnect tree? If it was not submitted upstream, we'll need to keep making sure to always fix this as we move to newer wiseconnect versions. Btw, the code comment above this line still says "Extention" :)

silabs-TiborL replied:
It's a pure downstream change, we could raise a ticket to WiseConnect team to fix this.

This typo is only present in the validity bits (which is currently the most significant bit).
So another possible solution would be to create new defines in the Zephyr repo with "BIT(31)". However if the SDK changes this can cause a "silent error" whereas the typo fix will result in a build error.

What would be your preference?

jhedberg replied:
We want to automate the importing of Silabs SDKs into `hal_silabs`. @asmellby has already done the initial work for the SiSDK in his latest PR. I think the cleanest and most sustainable option would be to raise a ticket to get this fixed upstream, since otherwise we would need to have extra logic in our automation to do the fix there. The option of having a custom define in Zephyr is the least preferred one, since we want to stay in sync with the upstream SDK and its API.

silabs-TiborL replied:
Okay, I'll leave it as is now and create a ticket about this.

jhedberg commented:
@asmellby @jerome-pouiller could we get a second review here? Thanks!

jerome-pouiller commented on wiseconnect/CMakeLists.txt:
```diff
@@ -61,44 +97,26 @@ zephyr_include_directories(
-zephyr_compile_definitions(
-    SLI_SI91X_OFFLOAD_NETWORK_STACK
-    SLI_SI91X_SOCKETS
+zephyr_compile_definitions_ifdef(CONFIG_SIWX917_NWP_NETWORK_STACK
```
Do we want to name this option `CONFIG_SIWX917_NWP_NETWORK_STACK` or `CONFIG_WISECONNECT_NETWORK_STACK`?

jhedberg replied:
Currently these are defined in `soc/silabs/silabs_siwx917/siwg917/Kconfig` in the main tree. If we change the name to `WISECONNECT_*` I think a more appropriate place would be `modules/Kconfig.silabs` (also in the main tree).

jerome-pouiller replied:
Yes, we can consider this option is more a "mode of operation" of WiseConnect rather than just a support for a driver. I feel it make sense to hide it in `modules/Kconfig.silabs` (note this file is currently named `modules/Kconfig` in our downstream)

jhedberg replied:
> Yes, we can consider this option is more a "mode of operation" of WiseConnect rather than just a support for a driver. I feel it make sense to hide it in `modules/Kconfig.silabs`

Sure, makes sense. Unrelated to the current case, but I have a feeling we may want to move to using `kconfig-ext` in `module.yml` and then use `modules/hal_silabs/Kconfig` in the main tree. That seems to be the more "modern" approach.

> note this file is currently named modules/Kconfig in our downstream

Hmm? I can't find a modules directory at all in zephyr-silabs.

silabs-TiborL replied:
Is it ok to do just the rename here and do the relocation in another PR?

jhedberg replied:
> Is it ok to do just the rename here and do the relocation in another PR?

So you agree this can be seen as a more WiseConnect specific thing rather than a 917 specific thing? I'm not opposed to just doing the rename at first, though in both cases you will anyway need to update your zephyr-silabs PR, i.e. it's not really that much additional work to create a `modules/Kconfig.silabs` that will eventually get merged with the corresponding file in the main tree.

Btw, I realized that we have `config HAS_SILABS_WISECONNECT` in the zephyr-silabs root Kconfig, which should really also be in the module Kconfig file.

jerome-pouiller replied:
>> note this file is currently named modules/Kconfig in our downstream
>
> [...]
>
> Btw, I realized that we have config HAS_SILABS_WISECONNECT in the zephyr-silabs root  Kconfig, which should really also be in the module Kconfig file.

Yes, it was what I wanted to say.

silabs-TiborL replied:
Yes, completely agree. 
You're right, I need to update both PR so then I'll do these Kconfig modifications as well.

jhedberg replied:
@silabs-TiborL I think there was some misunderstanding. The place I was suggesting to put this was in the `zephyr-silabs` tree and not the `hal_silabs` tree. The goal is also to remove all CMake and Kconfig files completely from `hal_silabs` and have those instead in the main tree (or `zephyr-silabs` in the case of pre-upstream 917 work).

For now I'd create a `modules/Kconfig.silabs` file in the `zephyr-silabs` tree and put all of this there. Please also see my comment in the #zephyr-development Slack channel how this should change in the long-term (however for now `modules/Kconfig.silabs` is the way to go, since that mirrors what's in the upstream main tree).

silabs-TiborL replied:
I see, sorry about this, I've only started working with Zephyr recently andI'm not 100% sure that I got your intention correctly. Thus I'd go for the easy path and just rename CONFIG_SIWX917_NWP_NETWORK_STACK to CONFIG_WISECONNECT_NETWORK_STACK for this PR.

jhedberg replied:
Sure. Let's go with that then in the short-term.

silabs-TiborL committed:
wiseconnect: CONFIG_SIWX917_NWP_NETWORK_STACK rename to CONFIG_WISECONNECT_NETWORK_STACK

Signed-off-by: Tibor Laczko <tibor.laczko@silabs.com>

jhedberg commented:
@jerome-pouiller are you ok with the intermediate solution? As a follow-up PR we can reorganize the Kconfig options and files in zephyr-silabs.
